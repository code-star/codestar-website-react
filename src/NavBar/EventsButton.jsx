import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import compose from 'recompose/compose';
import PropTypes from 'prop-types';

import { Button } from '@material-ui/core';
import { Notifications as NotificationImportantIcon } from '@material-ui/icons';
import { withStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import { jsonp } from '../util';

const styles = theme => ({
	newEventIcon: {
		color: 'red',
		marginRight: theme.spacing.unit,
	},
	newEventIconHover: {
		color: 'inherit',
	},
});

// Meetup API test console: https://secure.meetup.com/meetup_api/console/?path=/:urlname/events
// page=3 = number of results to return in a page, only need the first 3 results
const GET_UPCOMING_EVENTS_URL =
	'https://api.meetup.com/Code-Star-Night/events?photo-host=secure&page=3&sig_id=226887185&status=upcoming&sig=e3efc6db037cf681181d84ae343459a36afbefd4';

export class EventsButton extends Component {
	constructor(props) {
		super(props);
		this.state = {
			nextEvent: '',
			isHovering: false,
		};
		this.fetchUpcomingEvent = this.fetchUpcomingEvent.bind(this);
		this.handleMouseOver = this.handleMouseOver.bind(this);
		this.handleMouseOut = this.handleMouseOut.bind(this);
		this.fetchUpcomingEvent();
	}

	fetchUpcomingEvent() {
		/* Meetup API only allows JSONP for client-side, non authenticated, api key signed GET requests.
		   must use JSONP conform https://github.com/meetup/api/issues/211
		   Fetch API does not support JSONP. no-cors mode creates an opaque response without data.
		*/
		jsonp(GET_UPCOMING_EVENTS_URL).then(data => {
			const nextEvent = data.data[0];
			if (nextEvent) {
				const formattedDate = new Date(nextEvent.time).toLocaleDateString(
					'en-US',
					{ month: 'long', day: 'numeric' }
				);
				const text = `${formattedDate} *Meetup* ${nextEvent.name}`;
				this.setState({ nextEvent: text });
			}
		});
	}

	handleMouseOver() {
		this.setState({ isHovering: true });
	}

	handleMouseOut() {
		this.setState({ isHovering: false });
	}

	render() {
		const iconClasses = `${this.props.classes.newEventIcon} ${
			this.state.isHovering ? this.props.classes.newEventIconHover : null
		}`;
		const icon =
			this.state.nextEvent !== '' ? (
				<NotificationImportantIcon className={iconClasses} />
			) : null;
		return (
			<Tooltip title={this.state.nextEvent} placement="bottom">
				<Button
					component={Link}
					to="/events"
					color="inherit"
					onMouseOver={this.handleMouseOver}
					onMouseOut={this.handleMouseOut}
				>
					{icon}
					{this.props.label}
				</Button>
			</Tooltip>
		);
	}
}

EventsButton.propTypes = {
	label: PropTypes.string.isRequired,
};

export default compose(withStyles(styles))(EventsButton);
