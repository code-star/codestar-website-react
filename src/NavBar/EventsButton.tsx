import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import compose from 'recompose/compose';

import { Button } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import i18n from '../i18n';

// Fixme: this is a workaround for using the material ui button
// with the `to` property. By default this is not supported.
const CustomButton = (props: any) => <Button {...props} />;

// TODO improve types by replacing "any"
interface IEventsButtonPropsInner {
	classes: Record<string, string>;
	label: string;
	nextEvent: any;
}

interface IEventsButtonPropsOuter {
	label: string;
	nextEvent: any;
}

interface IEventsButtonState {
	isHovering: boolean;
}

const styles = (theme: any) => ({
	newEventIcon: {
		color: 'red',
		display: 'inline-block',
		fontSize: '1.3rem',
		marginLeft: theme.spacing.unit * 0.5,
	},
	newEventIconHover: {
		color: 'inherit',
	},
	bigTooltip: {
		fontSize: 20,
	},
});

// TODO to stateless with https://jsfiddle.net/evenchange4/p3vsmrvo/1599/
export class EventsButton extends Component<
	IEventsButtonPropsInner,
	IEventsButtonState
> {
	constructor(props: IEventsButtonPropsInner) {
		super(props);
		this.state = {
			isHovering: false,
		};
		this.handleMouseOver = this.handleMouseOver.bind(this);
		this.handleMouseOut = this.handleMouseOut.bind(this);
	}

	public handleMouseOver() {
		this.setState({ isHovering: true });
	}

	public handleMouseOut() {
		this.setState({ isHovering: false });
	}

	public render() {
		const { classes, label } = this.props;
		const iconClasses = `${classes.newEventIcon} ${
			this.state.isHovering ? classes.newEventIconHover : null
		}`;
		let icon = null;
		let nextEventText = '';
		if (this.props.nextEvent) {
			icon = <span className={iconClasses}> ‚óè</span>;
			const locale = i18n.language === 'nl' ? 'nl-NL' : 'en-US';
			const formattedDate = new Date(
				this.props.nextEvent.time
			).toLocaleDateString(locale, { month: 'long', day: 'numeric' });
			nextEventText = `${formattedDate} *Meetup* ${this.props.nextEvent.name}`;
		}
		return (
			<Tooltip
				title={nextEventText}
				placement="bottom"
				classes={{
					tooltip: classes.bigTooltip,
				}}
			>
				<CustomButton
					component={Link}
					to="/events"
					color="inherit"
					onMouseOver={this.handleMouseOver}
					onMouseOut={this.handleMouseOut}
				>
					{label}
					{icon}
				</CustomButton>
			</Tooltip>
		);
	}
}

export default compose<IEventsButtonPropsInner, IEventsButtonPropsOuter>(
	withStyles(styles)
)(EventsButton);
