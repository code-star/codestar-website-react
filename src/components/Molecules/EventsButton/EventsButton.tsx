import React, { SFC } from 'react';
import { Link } from 'react-router-dom';
import compose from 'recompose/compose';
import { withStateHandlers } from 'recompose';

import { Button } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';
import i18n from '../../../i18n';

// Fixme: this is a workaround for using the material ui button
// with the `to` property. By default this is not supported.
const CustomButton = (props: any) => <Button {...props} />;

// TODO improve types by replacing "any"
interface IEventsButtonPropsInner {
  classes: Record<string, string>;
  label: string;
  nextEvent: any;
  isHovering: boolean;
  handleMouseOver: any;
  handleMouseOut: any;
}

interface IEventsButtonPropsOuter {
  label: string;
  nextEvent: any;
}

const styles = (theme: any) => ({
  newEventIcon: {
    color: 'red',
    display: 'inline-block',
    fontSize: '1.3rem',
    marginLeft: theme.spacing.unit * 0.5,
  },
  newEventIconHover: {
    color: 'inherit',
  },
  bigTooltip: {
    fontSize: 20,
  },
});

export const EventsButton: SFC<IEventsButtonPropsInner> = ({
  classes,
  label,
  nextEvent,
  isHovering,
  handleMouseOver,
  handleMouseOut,
}) => {
  const iconClasses = `${classes.newEventIcon} ${
    isHovering ? classes.newEventIconHover : null
  }`;
  let icon = null;
  let nextEventText = '';
  if (nextEvent) {
    icon = <span className={iconClasses}> ‚óè</span>;
    const locale = i18n.language === 'nl' ? 'nl-NL' : 'en-US';
    const formattedDate = new Date(nextEvent.time).toLocaleDateString(locale, {
      month: 'long',
      day: 'numeric',
    });
    nextEventText = `${formattedDate} üéâMeetupüéâ ${nextEvent.name}`;
  }
  return (
    <Tooltip
      title={nextEventText}
      placement="bottom"
      classes={{
        tooltip: classes.bigTooltip,
      }}
    >
      <CustomButton
        component={Link}
        to="/events"
        color="inherit"
        onMouseOver={handleMouseOver}
        onMouseOut={handleMouseOut}
      >
        {label}
        {icon}
      </CustomButton>
    </Tooltip>
  );
};

export const initialState = { isHovering: false };
export const stateUpdaters = {
  // Example of state argument with typing: handleMouseOver: ({ isHovering } : { isHovering: boolean}) => () => ({ isHovering: true }),
  handleMouseOver: () => () => ({ isHovering: true }),
  handleMouseOut: () => () => ({ isHovering: false }),
};

export default compose<IEventsButtonPropsInner, IEventsButtonPropsOuter>(
  withStateHandlers(initialState, stateUpdaters),
  withStyles(styles)
)(EventsButton);
