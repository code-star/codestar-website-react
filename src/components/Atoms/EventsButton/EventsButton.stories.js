import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';
import { MemoryRouter } from 'react-router-dom';
import EventsButton from './EventsButton';

const stories = storiesOf('components/Atoms/EventsButton', module);

stories.addDecorator(withKnobs);

const label = 'Button label';
const defaultValue = 'Events';
const groupId = 'GROUP-ID1';

const value = text(label, defaultValue, groupId);

stories.add('with upcoming event', () => (
  <MemoryRouter>
    <EventsButton
      label={value}
      nextEvent={{
        time: 1540643542266,
        name: 'test',
      }}
    />
  </MemoryRouter>
));
