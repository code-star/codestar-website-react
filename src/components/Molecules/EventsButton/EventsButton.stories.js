import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, text } from '@storybook/addon-knobs'
import { MemoryRouter } from 'react-router-dom'
import EventsButton from './EventsButton'

const stories = storiesOf('components/Molecules/EventsButton', module)

stories.addDecorator(withKnobs)

stories.add('with upcoming event', () => {
  const labelValue = text('Button label', 'Events')
  const nextEventNameValue = text('Next event name', 'My Event')
  const nextEventDateValue = text('Next event date', new Date(1540643542266))
  const nextEventValue = {
    time: nextEventDateValue,
    name: nextEventNameValue,
  }
  return (
    <MemoryRouter>
      <EventsButton label={labelValue} nextEvent={nextEventValue} />
    </MemoryRouter>
  )
})
