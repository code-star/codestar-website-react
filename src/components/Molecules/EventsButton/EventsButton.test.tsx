import React from 'react';
import ReactDOM from 'react-dom';
import { EventsButton, initialState, stateUpdaters } from './EventsButton';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

jest.mock('../../../i18n', () => ({
  language: 'nl',
}));

const handleMouseOut = () => '';
const handleMouseOver = () => '';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <MemoryRouter>
      <EventsButton
        label="TEST_EVENTS_LABEL"
        classes={{ newEventIcon: '' }}
        nextEvent={null}
        isHovering={false}
        handleMouseOut={handleMouseOut}
        handleMouseOver={handleMouseOver}
      />
    </MemoryRouter>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

describe('An instance of EventsButton', () => {
  it('matches the snapshot without Event', () => {
    const testRenderer = renderer.create(
      <MemoryRouter>
        <EventsButton
          classes={{ newEventIcon: '' }}
          label="TEST_EVENTS_LABEL_2"
          nextEvent={null}
          isHovering={false}
          handleMouseOut={handleMouseOut}
          handleMouseOver={handleMouseOver}
        />
      </MemoryRouter>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('matches the snapshot while hovering', () => {
    const testRenderer = renderer.create(
      <MemoryRouter>
        <EventsButton
          classes={{ newEventIcon: '' }}
          label="TEST_EVENTS_LABEL_2"
          nextEvent={null}
          isHovering={true}
          handleMouseOut={handleMouseOut}
          handleMouseOver={handleMouseOver}
        />
      </MemoryRouter>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('matches the snapshot with Event', () => {
    const testRenderer = renderer.create(
      <MemoryRouter>
        <EventsButton
          classes={{ newEventIcon: '' }}
          label="TEST_EVENTS_LABEL_2"
          nextEvent={{ name: 'MyEvent', time: 1535391830804 }}
          isHovering={false}
          handleMouseOut={handleMouseOut}
          handleMouseOver={handleMouseOver}
        />
      </MemoryRouter>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  // TODO improvement would be to test if the DOM mouseOver of CustomButton triggers the (mocked) function handleMouseOver

  it('has an initial hover state', () => {
    expect(initialState).toEqual({
      isHovering: false,
    });
  });

  it('sets hover state', () => {
    expect(stateUpdaters.handleMouseOver()()).toEqual({
      isHovering: true,
    });
  });

  it('unsets hover state', () => {
    expect(stateUpdaters.handleMouseOut()()).toEqual({
      isHovering: false,
    });
  });
});
