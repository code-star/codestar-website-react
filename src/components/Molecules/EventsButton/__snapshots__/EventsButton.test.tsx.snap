// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An instance of EventsButton matches the snapshot while hovering 1`] = `
<a
  aria-describedby={null}
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-flat MuiButton-colorInherit"
  href="/events"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
  title=""
>
  <span
    className="MuiButton-label"
  >
    TEST_EVENTS_LABEL_2
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</a>
`;

exports[`An instance of EventsButton matches the snapshot with Event 1`] = `
<a
  aria-describedby={null}
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-flat MuiButton-colorInherit"
  href="/events"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
  title="August 27 üéâMeetupüéâ MyEvent"
>
  <span
    className="MuiButton-label"
  >
    TEST_EVENTS_LABEL_2
    <span
      className=" null"
    >
       ‚óè
    </span>
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</a>
`;

exports[`An instance of EventsButton matches the snapshot without Event 1`] = `
<a
  aria-describedby={null}
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-flat MuiButton-colorInherit"
  href="/events"
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
  title=""
>
  <span
    className="MuiButton-label"
  >
    TEST_EVENTS_LABEL_2
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</a>
`;
